rhombus/private/list.rkt

```diff
@@ -289,64 +291,72 @@
-  v)
-
-(define-static-info-syntax list
-  (#%call-result #,list-static-infos)
-  (#%function-arity -1))
+(define (check-repetition-list who l)
+  (check-list who l)
+  l)
 
-(define-static-info-syntax iota
-  (#%call-result #,list-static-infos)
-  (#%function-arity 2))
+(define (check-function-of-arity n who proc)
+  (unless (and (procedure? proc)
+               (procedure-arity-includes? proc n))
+    (raise-argument-error* who rhombus-realm
+                           (string-append "Function.of_arity("
+                                          (number->string n)
+                                          ")")
+                           proc)))
 
-(define/arity (List.map lst proc)
+(define/method (List.map lst proc)
   #:static-infos ((#%call-result #,list-static-infos))
-  (unless (procedure? proc)
-    (raise-argument-error* 'List.map rhombus-realm "Function" proc))
+  (check-list who lst)
+  (check-function-of-arity 1 who proc)
   (map proc lst))
 
-(define/arity (List.for_each lst proc)
-  (unless (procedure? proc)
-    (raise-argument-error* 'List.for_each rhombus-realm "Function" proc))
+(define/method (List.for_each lst proc)
+  (check-list who lst)
+  (check-function-of-arity 1 who proc)
   (for-each proc lst))
 
-(define/arity (List.sort lst [less-than? <])
+(define/method (List.sort lst [less-than? <])
   #:static-infos ((#%call-result #,list-static-infos))
-  (unless (and (procedure? less-than?)
-               (procedure-arity-includes? less-than? 2))
-    (raise-argument-error* 'List.sort rhombus-realm "Function.of_arity(2)" less-than?))
+  (check-list who lst)
+  (check-function-of-arity 2 who less-than?)
   (sort lst less-than?))
 
-(define/arity List.append
+;; error adjusting in "runtime-config.rkt"
+;; only check that the *last* argument is list here
+(define/method List.append
   #:static-infos ((#%call-result #,list-static-infos))
   (case-lambda
     [() null]
     [(a)
-     (unless (list? a) (raise-argument-error* 'List.append rhombus-realm "List" a))
+     (check-list who a)
      a]
     [(a b)
-     (cond
-       [(list? b) (append a b)]
-       [else
-        (unless (list? a) (raise-argument-error* 'List.append rhombus-realm "List" a))
-        (raise-argument-error* 'List.append rhombus-realm "List" b)])]
+     (unless (list? b)
+       (check-list who a)
+       (raise-argument-error* who rhombus-realm "List" b))
+     (append a b)]
     [ls
-     (if (list? (let loop ([ls ls]) (if (pair? (cdr ls)) (loop (cdr ls)) (car ls))))
-         (apply append ls)
+     (let ([ln (let loop ([ls ls])
+                 (if (pair? (cdr ls)) (loop (cdr ls)) (car ls)))])
+       (unless (list? ln)
          (for ([l (in-list ls)])
-           (unless (list? l) (raise-argument-error* 'List.append rhombus-realm "List" l))))]))
+           (check-list who l))
+         (raise-argument-error* who rhombus-realm "List" ln)))
+     (apply append ls)]))
+
+;; error adjusting in "runtime-config.rkt"
```

Comment:
For some reason, `List.get` doesn't seem to get inlined as `list-ref` by schemify, so that's worth a closer look. Also, we may want to have an exposed `List.get` operation that checks for listness (not just a sequence of pairs), and then distinction between `list-ref` for `#%index` and `List.get` may be worthwhile; I'm not sure.